README.md
# VERIFY
[![GoDoc](https://godoc.org/github.com/pirmd/verify?status.svg)](https://godoc.org/github.com/pirmd/verify)&nbsp; 
[![Go Report Card](https://goreportcard.com/badge/github.com/pirmd/verify)](https://goreportcard.com/report/github.com/pirmd/verify)&nbsp;

`verify` package is a collection of simple functions that I find handy to use
for building my unit testing.

# INSTALLATION
Everything should work fine using go standard commands (`build`, `get`,
`install`...).

# USAGE
Running `go doc github.com/pirmd/verify` should give you helpful guidelines on
availables features.

With the package set-up, additional go test flags are offered:
    . `-test.golden-update`: updates the golden files with the test result.
    . `-test.mockhttp-update`: updates files served through the mock-http transport.
    . `-test.diff`: show differences between the test result and expected values.
    . `-test.diff-color`: show differences in color between the test result and the
       expected values.
    . `-test.diff-np`: show differences between result and expected values
      materializing non printable chars.

# CONTRIBUTION
If you feel like to contribute, just follow github guidelines on
[forking](https://help.github.com/articles/fork-a-repo/) then [send a pull
request](https://help.github.com/articles/creating-a-pull-request/)

[modeline]: # ( vim: set fenc=utf-8 spell spl=en: )


CHANGELOG.md
# Changelog
## [0.8.0] - 2022-04-08 
- Allow golden files and mocked HTTP responses locations to be accessed
  and modified by end-user.

## [0.7.0] - 2022-03-20
- Simplify files-related testing by taking benefit of new
  testing.T.TempDir()

## [0.6.0] - 2021-03-14
- Update dependencies to github.com/pirmd/text v0.6.0
- Get rid of external dependency (github.com/sanity-io/litter).

## [0.5.3] - 2020-12-01
### Modified
- Get rid of external dependency (github.com/mattetti/filebuffer).

## [0.5.2] - 2020-11-29
### Modified
- Extend MockROFile to support io.ReaderAt and io.Seeker interfaces.

## [0.5.1] - 2020-07-12
### Added
- Add function to verify that a folder is empty

## [0.5.0] - 2020-05-16
### Added
- Simple wrapper around testing.TB to allow plumbing testing log facilities to
  an existing logger.

## [O.4.2] - 2020-05-08 
### Modified
- FIX diff without color to show empty right text when an insertion is detected.
- Improve support/feedback to user in case golden files do not exist.

## [0.4.1] - 2020-02-17
### Modified
- FIX an unwanted behavior following introduction of
  github.com/sanity-io/litter as stringifier where all strings where quoted
  using strconv.Quote. It is not the behavior I expect in most cases, bypass it
  as a short term workaround.

## [0.4.0] - 2020-02-16
### Modified
- Break previous principles for verify functions. Verify functions are now
  outputting errors in case of check fail and leave the end user the choice on
  how to handle the situation in its unit testing code (for example t.Errorf or
  t.Fatalf...)
- Switch to github.com/pirmd/text v0.4.0 that offers better diff algorithm and
  color diff formatting.
- Use github.com/sanity-io/litter to provide readily usable pretty formatting
  of compared values. 


## [0.3.0] - 2019-11-11
### Added
- Add extended error message definition in MatchGolden
- Add function to list all files matching a pattern from a TestField
- Add ListWithExt function to TestField
- Add helpers to capture os.Stdout and compare it to either a string or a
  golden file
- Add basic mock http server that alters http.DefaultTransport and replace it
  to serve locally stored http.Response
### Modified
- Change comparison functions to always expect 'got' and 'want' arguments in
  the same order


## [0.2.0] - 2019-08-11
### Added
- Add a flag to displays differences in colors
### Modified
- Correct a BUG when displaying diff in test results
- Modified slightly the formatting of error messages in case of tests' errors 

## [0.1.0] - 2019-05-11
### Added
- basic comparison with optional diff output between 'expected' and 'got' result
- support of golden files and there management (updating)
- basic function to play with temporary testing files


file.go
package verify

import (
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// MockReader is an interface that offers most of os.File reading operations.
type MockReader interface {
	io.Reader
	io.ReaderAt
	io.Seeker
}

// MockROFile returns a bytes.Buffer backed mock File that implements io.Reader,
// io.Seeker, io.ReaderAt interfaces.
func MockROFile(content string) MockReader {
	return strings.NewReader(content)
}

// FileExists checks if provided path exists.
func FileExists(path string) error {
	_, err := os.Stat(path)
	if os.IsNotExist(err) {
		return errors.New(path + " does not exist")
	}
	if err != nil {
		return fmt.Errorf("cannot stat '%s': %s", path, err)
	}
	return nil
}

// FileDoesNotExist checks if provided path does not exist.
func FileDoesNotExist(path string) error {
	_, err := os.Stat(path)
	if os.IsNotExist(err) {
		return nil
	}
	if err != nil {
		return fmt.Errorf("cannot stat '%s': %s", path, err)
	}
	return errors.New(path + " does exist (but should not)")
}

// DirHasContent checks that the given dir contains the provided tree. Tree
// should be relative to root.
func DirHasContent(root string, tree []string) error {
	ls, err := lsDir(root)
	if err != nil {
		return err
	}

	return EqualSliceWithoutOrder(ls, tree)
}

// DirIsEmpty checks that a dir is empty.
func DirIsEmpty(path string) error {
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	if _, err = f.Readdir(1); err != io.EOF {
		return errors.New(path + " is not empty (but should be)")
	}
	return nil
}

// TestFolder represents a temporary folder where testing can happen
type TestFolder struct {
	Root string
}

// MustNewTestFolder creates a temporary folder to host a test folder for tb.
// If creation fails, tb is terminated by calling tb.Fatal.
func MustNewTestFolder(tb testing.TB) *TestFolder {
	return &TestFolder{
		Root: tb.TempDir(),
	}
}

// Populate populates a temporary testing folders with the given tree The
// provided list of files should be in the correct order where folders are
// listed before the files they contain (if any). Any provided path is relative
// to the temporary testing folder root. Any path provided without any extension
// is interpreted as being a folder.
func (tmp *TestFolder) Populate(tree []string) error {
	for _, f := range tree {
		path := tmp.Fullpath(f)
		if filepath.Ext(path) == "" {
			if err := os.MkdirAll(path, 0777); err != nil {
				return fmt.Errorf("fail to create temporary folder (path %s): %v", f, err)
			}
		} else {
			if err := os.MkdirAll(filepath.Dir(path), 0777); err != nil {
				return fmt.Errorf("fail to create file %s: %v", path, err)
			}

			if _, err := os.Create(path); err != nil {
				return fmt.Errorf("fail to create temporary file (path %s): %v", f, err)
			}
		}
	}

	return nil
}

// List returns the list of files and folders contained in the temporary test
// folder. Returned tree is made of relative path to testing folder's root. Order
// is the lexical order (as it uses filepath.Walk under the hood).
func (tmp *TestFolder) List() ([]string, error) {
	ls, err := lsDir(tmp.Root)
	if err != nil {
		return nil, fmt.Errorf("fail to list temporary test folder: %v", err)
	}

	return ls, nil
}

// Glob returns the list of files and folders contained in the temporary test
// folder that match the given pattern. Returned tree is made of relative path
// to testing folder's root. Order is the lexical order (as it uses
// filepath.Walk under the hood).
func (tmp *TestFolder) Glob(pattern string) (tree []string, err error) {
	err = filepath.Walk(tmp.Root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if path == tmp.Root {
			return nil
		}

		relpath, _ := filepath.Rel(tmp.Root, path) //Sure that we are not going to fail here

		match, err := filepath.Match(pattern, relpath)
		if err != nil {
			return err
		}
		if !match {
			return nil
		}

		tree = append(tree, relpath)
		return nil
	})

	if err != nil {
		return nil, fmt.Errorf("fail to list temporary test folder (path %s): %v", tmp.Root, err)
	}

	return
}

// ListWithExt returns the list of files contained in the temporary test folder
// whose extension match the given pattern.  Returned tree is made of relative
// path to testing folder's root. Order is the lexical order (as it uses
// filepath.Walk under the hood).
func (tmp *TestFolder) ListWithExt(ext string) (tree []string, err error) {
	err = filepath.Walk(tmp.Root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if path == tmp.Root {
			return nil
		}

		if filepath.Ext(path) != ext {
			return nil
		}

		relpath, _ := filepath.Rel(tmp.Root, path) //Sure that we are not going to fail here
		tree = append(tree, relpath)
		return nil
	})

	if err != nil {
		return nil, fmt.Errorf("fail to list temporary test folder (path %s): %v", tmp.Root, err)
	}

	return
}

// Fullpath returns the complete path to the given relative path from the test
// folder root
func (tmp *TestFolder) Fullpath(relpath string) string {
	return filepath.Join(tmp.Root, filepath.Clean("/"+relpath))
}

// ShouldHaveFile check if provided path exists in the test folder. The path can be
// relative to the test folder's root.
func (tmp *TestFolder) ShouldHaveFile(relpath string) error {
	path := tmp.Fullpath(relpath)
	return FileExists(path)
}

// ShouldNotHaveFile checks that the given files are not in the test folder
func (tmp *TestFolder) ShouldNotHaveFile(wanted string) error {
	path := tmp.Fullpath(wanted)
	return FileDoesNotExist(path)
}

// ShouldHaveContent verifies that the temporary test folder content corresponds
// to the wanted tree.
func (tmp *TestFolder) ShouldHaveContent(wanted []string) error {
	return DirHasContent(tmp.Root, wanted)
}

// ShouldBeEmpty verifies that the temporary test folder is empty.
func (tmp *TestFolder) ShouldBeEmpty() error {
	return DirIsEmpty(tmp.Root)
}

func lsDir(root string) (tree []string, err error) {
	err = filepath.Walk(root, func(path string, fi os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if path == root {
			return nil
		}
		relpath, _ := filepath.Rel(root, path) //Sure that we are not going to fail here
		if fi.IsDir() {
			relpath = relpath + string(os.PathSeparator)
		}

		tree = append(tree, relpath)
		return nil
	})

	return
}


http.go
package verify

import (
	"bufio"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/http/httputil"
	"os"
	"path/filepath"
)

var (
	// updateMockHTTP updates the http.Response files provided through MockHTTPResponse
	updateMockHTTP = flag.Bool("test.mockhttp-update", false, "update files served through the mock http transport")
)

// MockHTTPResponse represents a mock http transport for testing purpose. It
// implements http.RoundTripper, which handles single http requests issued by a
// http.Client.
//
// MockHTTPResponse returns the content of a file from mockHTTPDir whose
// filename is the requested url. If no file is found with requested url as
// filename, an empty http.Response if no corresponding file exists.
type MockHTTPResponse struct {
	mockHTTPDir   string
	origTransport http.RoundTripper
}

// NewMockHTTPResponse creates a new MockHTTPResponse returning content stored
// in mockHTTPDir.
func NewMockHTTPResponse(mockHTTPDir string) *MockHTTPResponse {
	return &MockHTTPResponse{
		mockHTTPDir: mockHTTPDir,
	}
}

// StartMockHTTPResponse creates a new MockHTTPResponse and starts it.
func StartMockHTTPResponse(mockHTTPDir string) *MockHTTPResponse {
	m := NewMockHTTPResponse(mockHTTPDir)
	m.Start()
	return m
}

// Start actually starts the MockHTTP transport by replacing
// http.DefaultTransport by itself.
func (m *MockHTTPResponse) Start() {
	m.origTransport = http.DefaultTransport
	http.DefaultTransport = m
}

// Stop stops MockHTTPResponse and restore initial http.DefaultTransport.
func (m *MockHTTPResponse) Stop() {
	http.DefaultTransport = m.origTransport
}

// RoundTrip implements http.Roundtripper interface.
func (m *MockHTTPResponse) RoundTrip(req *http.Request) (*http.Response, error) {
	if *updateMockHTTP {
		if err := m.updateResponsesFiles(req); err != nil {
			return nil, fmt.Errorf("fail to update mock http response file (in %s): %v", m.pathFor(req), err)
		}
	}

	resp, err := m.respondFromFiles(req)
	if err != nil {
		return nil, fmt.Errorf("fail to mock http response (for %s): %v", req.URL, err)
	}

	return resp, err
}

func (m *MockHTTPResponse) respondFromFiles(req *http.Request) (*http.Response, error) {
	f, err := os.Open(m.pathFor(req))
	if err != nil {
		return nil, err
	}

	resp, err := http.ReadResponse(bufio.NewReader(f), nil)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

func (m *MockHTTPResponse) updateResponsesFiles(req *http.Request) error {
	m.Stop()

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}

	b, err := httputil.DumpResponse(resp, true)
	if err != nil {
		return err
	}

	path := m.pathFor(req)
	if err := os.MkdirAll(filepath.Dir(path), 0777); err != nil {
		return err
	}

	if err := ioutil.WriteFile(path, b, 0666); err != nil {
		return err
	}

	m.Start()

	return nil
}

func (m *MockHTTPResponse) pathFor(req *http.Request) string {
	return filepath.Join(m.mockHTTPDir, req.URL.Host, req.URL.Path, req.URL.RawQuery+".http")
}


stdout.go
package verify

import (
	"bytes"
	"fmt"
	"io"
	"os"
)

// MockStdout is an helper to capture output to Stdout.
type MockStdout struct {
	orig     *os.File
	r, w     *os.File
	captured chan string
}

// NewMockStdout creates a new record session of os.Sdtout and start capture
// operation.
func NewMockStdout() *MockStdout {
	return &MockStdout{
		orig:     os.Stdout,
		captured: make(chan string),
	}
}

// StartMockStdout creates a new record session of os.Sdtout and start capture
// operation.
func StartMockStdout() (*MockStdout, error) {
	out := NewMockStdout()
	if err := out.Start(); err != nil {
		return nil, err
	}
	return out, nil
}

// Start starts capturing os.Stdout.
func (out *MockStdout) Start() error {
	var err error

	out.r, out.w, err = os.Pipe()
	if err != nil {
		return fmt.Errorf("fail to capture Stdout: %v", err)
	}

	os.Stdout = out.w

	go func() {
		var buf bytes.Buffer
		io.Copy(&buf, out.r)
		out.captured <- buf.String()
	}()

	return nil
}

// Stop terminates a os.Stdout's recording session and restore os.Stdout to its
// original state
func (out *MockStdout) Stop() {
	out.w.Close()
	os.Stdout = out.orig
}

// String stops the os.Stdout's recording and returns the content of os.Stdout
// capture session.
func (out *MockStdout) String() string {
	out.Stop()
	return <-out.captured
}

// EqualStdoutString verifies that captured os.Stdout is equal to 'want' and
// feedback a test error message with a line by line diff between them
func EqualStdoutString(got *MockStdout, want string) error {
	return Equal(got.String(), want)
}

// MatchStdoutGolden compares captured os.Stdout to the content of a 'golden'
// file If 'update' command flag is used, update the 'golden' file
func MatchStdoutGolden(name string, got *MockStdout) error {
	return MatchGolden(name, got.String())
}


verify.go
package verify

import (
	"flag"
	"fmt"

	"github.com/pirmd/text/diff"
	"github.com/pirmd/text/table"
)

var (
	showdiff      = flag.Bool("test.diff", false, "show differences between result and expected values")
	showdiffcolor = flag.Bool("test.diff-color", false, "show differences using colors between result and expected values")
	showdiffNP    = flag.Bool("test.diff-np", false, "show differences between result and expected values materializing non printable chars")
)

func msgWithDiff(got, want interface{}) string {
	g, w := stringify(got), stringify(want)

	if *showdiff || *showdiffcolor || *showdiffNP {
		delta := diff.Patience(w, g, diff.ByLines, diff.ByRunes)

		h := []diff.Highlighter{}
		if *showdiff {
			h = []diff.Highlighter{diff.WithSoftTabs, diff.WithoutMissingContent}
		}
		if *showdiffcolor {
			h = []diff.Highlighter{diff.WithSoftTabs, diff.WithColor}
		}
		if *showdiffNP {
			if !*showdiff && !*showdiffcolor {
				h = []diff.Highlighter{diff.WithSoftTabs, diff.WithoutMissingContent}
			}
			h = append([]diff.Highlighter{diff.WithNonPrintable}, h...)
		}

		dL, dR, dT, _ := delta.PrettyPrint(h...)
		return table.New().AddCol(dR, dT, dL).SetHeader("Got", "", "Want").String()
	}

	return fmt.Sprintf("Got:\n%v\n\nWant :\n%v", g, w)
}

func stringify(v interface{}) string {
	return fmt.Sprint(v)
}


golden.go
package verify

import (
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
)

var (
	//GoldenDir is the path where to find golden files
	GoldenDir = "./testdata"

	// updateGolden golden files, when set, updates the content of the golden
	// files from th etest result.
	updateGolden = flag.Bool("test.golden-update", false, "update golden file with test result")
)

// MatchGolden compares a test result to the content of a 'golden' file
// If 'update' command flag is used, update the 'golden' file
func MatchGolden(name string, got string) error {
	goldenPath := filepath.Join(GoldenDir, name+".golden")

	if *updateGolden {
		if err := updateGoldenFiles(goldenPath, []byte(got)); err != nil {
			return err
		}
	}

	want, err := readGolden(goldenPath)
	if err != nil {
		return fmt.Errorf("cannot read golden file %s: %s.\nTest output is:\n%s", goldenPath, err, got)
	}

	return Equal(got, string(want))
}

func readGolden(path string) ([]byte, error) {
	want, err := ioutil.ReadFile(path)
	if err != nil {
		return []byte{}, err
	}
	return want, nil
}

func updateGoldenFiles(path string, actual []byte) error {
	if err := os.MkdirAll(filepath.Dir(path), 0777); err != nil {
		return fmt.Errorf("cannot update golden file %s: %v", path, err)
	}

	if err := ioutil.WriteFile(path, actual, 0666); err != nil {
		return fmt.Errorf("cannot update golden file %s: %v", path, err)
	}

	return nil
}
